central/reports/scheduler/v2/schedule.go

```diff
@@ -0,0 +1,624 @@
+		AddExactMatches(search.ReportType, storage.ReportConfiguration_VULNERABILITY.String()).
+		ProtoQuery()
+	reportConfigs, err := s.reportConfigDatastore.GetReportConfigurations(ctx, query)
+	if err != nil {
+		log.Error("Error finding scheduled reports: %s", err)
+	}
+	for _, rc := range reportConfigs {
+		if rc.GetSchedule() != nil {
+			if err := s.UpsertReportSchedule(rc); err != nil {
+				log.Errorf("Error queuing scheduled report for report configuration with ID %s: %v", rc.GetId(), err)
+			}
+		}
+	}
+}
+
+/* Utility Functions */
+
+func (s *scheduler) doesUserHavePendingReport(ctx context.Context, configID string, userID string) (bool, error) {
+	s.dbLock.Lock()
+	defer s.dbLock.Unlock()
+
+	query := search.NewQueryBuilder().
+		AddExactMatches(search.ReportConfigID, configID).
+		AddExactMatches(search.ReportState, storage.ReportStatus_WAITING.String(), storage.ReportStatus_PREPARING.String()).
+		AddExactMatches(search.ReportRequestType, storage.ReportStatus_ON_DEMAND.String()).
+		ProtoQuery()
+	runningReports, err := s.reportMetadataStore.SearchReportMetadatas(ctx, query)
+	if err != nil {
+		return false, err
+	}
+	for _, rep := range runningReports {
+		if rep.GetRequester().GetId() == userID {
+			return true, nil
+		}
+	}
+	return false, nil
+}
+
+func (s *scheduler) lastSuccesfulScheduledReportTime(ctx context.Context, config *storage.ReportConfiguration) (*types.Timestamp, error) {
+	query := search.NewQueryBuilder().
+		AddExactMatches(search.ReportConfigID, config.GetId()).
+		AddExactMatches(search.ReportRequestType, storage.ReportStatus_SCHEDULED.String()).
+		AddExactMatches(search.ReportState, storage.ReportStatus_SUCCESS.String()).
+		WithPagination(search.NewPagination().
+			AddSortOption(search.NewSortOption(search.ReportCompletionTime).Reversed(true)).
+			Limit(1)).
+		ProtoQuery()
+	results, err := s.reportMetadataStore.SearchReportMetadatas(ctx, query)
+	if err != nil {
+		return nil, errors.Wrap(err, "Error finding last successful scheduled report time")
+	}
+	if len(results) > 1 {
+		return nil, errors.Errorf("Received %d records when only one record is expected", len(results))
+	}
+	if len(results) == 0 {
+		return nil, nil
+	}
+	return results[0].GetReportStatus().GetCompletedAt(), nil
+}
+
+func (s *scheduler) upsertReportStatus(ctx context.Context, metadata *storage.ReportMetadata, status storage.ReportStatus_RunState) error {
+	metadata.ReportStatus.RunState = status
+	return s.reportMetadataStore.UpdateReportMetadata(ctx, metadata)
+}
+
+func selectContext(req *ReportRequest) context.Context {
+	if req.Ctx == nil {
+		return scheduledCtx
+	}
+	return req.Ctx
+}
+
+func (s *scheduler) logAndUpsertError(reportErr error, req *ReportRequest) {
+	log.Errorf("Error while running report for config '%s': %s", req.ReportConfig.GetName(), reportErr)
+
+	req.ReportMetadata.ReportStatus.ErrorMsg = reportErr.Error()
+	req.ReportMetadata.ReportStatus.CompletedAt = types.TimestampNow()
```

Comment:
nil check `ReportMetadata` and `ReportStatus`

Reply:
There are lots of places in the implementation that read/write the report request. I think it would be best to add a validator and validate the request in `SubmitReport`. After that, we assume all helper functions have the valid request.